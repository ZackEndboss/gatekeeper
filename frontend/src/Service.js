import React from 'react';
import PropTypes from 'prop-types';
import R from 'ramda';
import {connect} from 'react-redux';
import {Col, Grid, Row, Glyphicon} from 'react-bootstrap';
import {openEditServiceDialog} from './reducers/dialog';
import Gate from './Gate';

export function Gates({environment_order, group, service, environments}) {
    return (
        <div key={`col-${group}-${service}`}>
            {R.map((group_environment) => {
                if (environments && environments.includes(group_environment)) {
                    return (
                        <Col xs={2} md={2} key={`col-${group}-${service}-${group_environment}`}>
                            <Gate key={`gate-${group}-${service}-${group_environment}`}
                                  group={group}
                                  service={service}
                                  environment={group_environment}/>
                        </Col>);
                } else {
                    return <Col xs={2} md={2} key={`col-${group}-${service}-${group_environment}`}/>;
                }
            })(environment_order)}
        </div>
    );
}

export class Service extends React.Component {
    render() {
        const {group, service, environments, environment_order, openEditServiceDialog} = this.props;
        return (
            <Grid fluid={true}>
                <Row>
                    <Col xs={1} md={1}>
                        {service}
                        <a className='glyp-button'
                           onClick={() => openEditServiceDialog()}>
                            <Glyphicon glyph='pencil'/>
                        </a>
                    </Col>
                    <Gates environment_order={environment_order} group={group} service={service}
                           environments={environments}/>
                </Row>
            </Grid>
        );
    }
}

Service.propTypes = {
    group: PropTypes.string.isRequired,
    service: PropTypes.string.isRequired,
    environment_order: PropTypes.array.isRequired,
    environments: PropTypes.array,
};

const defaultValues = {
    group: '',
    service: '',
    environment_order: [],
    environments: [],
};

const mapStateToProps = (state, initialProps) => {
    const gate_from_state = R.path(['gates', initialProps.group, initialProps.service])(state);
    const environments = {
        environments: Object.keys(gate_from_state)
    };
    return R.mergeAll([defaultValues, initialProps, environments]);
};

const mapDispatchToProps = (dispatch, initialProps) => {
    return {
        openEditServiceDialog: () => {
            dispatch(openEditServiceDialog(initialProps.group, initialProps.service));
        }
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Service);
