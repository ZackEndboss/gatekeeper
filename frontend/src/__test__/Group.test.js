import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import {MockStore} from './mockStore';
import {Provider} from 'react-redux';
import ConnectedGroup, {Group, Services} from '../Group';
import Service from '../Service';

jest.mock('../Service');

describe('Group component', () => {
    it('render', () => {
        const component = renderer.create(<Group group='group'
                                                 services={['service1', 'service2', 'service3']}/>);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should contain all services', () => {
        const component = shallow(<Services group='group'
                                            services={['service1', 'service2', 'service3']}/>);
        expect(component.find(Service).length).toEqual(3);
    });

});

describe('load from state', () => {
    it('should load the state correctly', () => {
        const storeMock = MockStore({
            'gates': {
                'ftx': {
                    'gatekeeper': {
                        'test1': {},
                        'test2': {}
                    }
                },
                'fty': {
                    'gatekeeper2': {
                        'test1': {},
                        'test2': {}
                    }
                }
            }
        });

        const subject = (
            <Provider store={storeMock}>
                <ConnectedGroup group='ftx'/>
            </Provider>);
        const component = renderer.create(subject);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});