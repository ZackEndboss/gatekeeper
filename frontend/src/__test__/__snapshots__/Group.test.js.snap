// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group component render 1`] = `
<div>
  <div
    className="container-fluid"
  >
    <div
      className="row"
    >
      <div
        className="col-md-1 col-xs-1"
      >
         
      </div>
      <div
        className="col-md-2 col-xs-2"
      >
        environment_1
      </div>
      <div
        className="col-md-2 col-xs-2"
      >
        environment_2
      </div>
      <div
        className="col-md-2 col-xs-2"
      >
        environment_3
      </div>
    </div>
  </div>
  <div>
    <div
      style={
        Object {
          "marginBottom": "20px",
        }
      }
    >
      <Service
        environment_order={
          Array [
            "environment_1",
            "environment_2",
            "environment_3",
          ]
        }
        group="group"
        service="service1"
      />
    </div>
    <div
      style={
        Object {
          "marginBottom": "20px",
        }
      }
    >
      <Service
        environment_order={
          Array [
            "environment_1",
            "environment_2",
            "environment_3",
          ]
        }
        group="group"
        service="service2"
      />
    </div>
    <div
      style={
        Object {
          "marginBottom": "20px",
        }
      }
    >
      <Service
        environment_order={
          Array [
            "environment_1",
            "environment_2",
            "environment_3",
          ]
        }
        group="group"
        service="service3"
      />
    </div>
  </div>
</div>
`;

exports[`load from state should load the state correctly 1`] = `
<div>
  <div
    className="container-fluid"
  >
    <div
      className="row"
    >
      <div
        className="col-md-1 col-xs-1"
      >
         
      </div>
      <div
        className="col-md-2 col-xs-2"
      >
        test1
      </div>
      <div
        className="col-md-2 col-xs-2"
      >
        test2
      </div>
    </div>
  </div>
  <div>
    <div
      style={
        Object {
          "marginBottom": "20px",
        }
      }
    >
      <Service
        environment_order={
          Array [
            "test1",
            "test2",
          ]
        }
        group="ftx"
        service="gatekeeper"
      />
    </div>
  </div>
</div>
`;
